### PR TITLE
Sdio/bioconda recipe parsing

### PR DESCRIPTION
**Pull request by spearw**

### CODE REVIEW

**Comment by jspeed-meyers** [comment]
@spearw, welcome to the repo and project! Before merging, please have @ralatsdc and myself review. Ray because he is so skilled and me just so I can keep a global picture of this codebase in my mind. Again, thank you and look forward to working with you.

**Comment by jspeed-meyers** [comment]
And feel free to rewrite tests and add test infrastructure as needed.

**Comment by spearw** [comment]
> @spearw, welcome to the repo and project! Before merging, please have @ralatsdc and myself review. Ray because he is so skilled and me just so I can keep a global picture of this codebase in my mind. Again, thank you and look forward to working with you.



> @spearw, welcome to the repo and project! Before merging, please have @ralatsdc and myself review. Ray because he is so skilled and me just so I can keep a global picture of this codebase in my mind. Again, thank you and look forward to working with you.

Excited to be collaborating on it! I'll be sure to include you both in all reviews. Feel free to ask if you have any questions about what I've done.

File: bioconda.py
Comment (ralatsdc) [inline bioconda.py:52]: Why is `clean_github_link` not needed here?

File: bioconda.py
Comment (ralatsdc) [inline bioconda.py:9]: Check the `filepath` documentation. I think it might need to be updated.

**Review by ralatsdc — CHANGES_REQUESTED** [review]
Please report the status of testing in the description of this pull request.

File: README.md
Comment (jspeed-meyers) [inline README.md:32]: @spearw, these are helpful instructions. Thank you! I do think they might overwhelm the user if they are in the main README though. Would you please create a separate, bioconda-specific README and link to that README. That way this main README is less cognitively demanding but has a link to the full explanation. Again, nice instructions!

File: bioconda.py
Comment (jspeed-meyers) [reply bioconda.py:52]: Same. And feel free to revise ```clean_github_link``` if it needs to be revised.

Comment (jspeed-meyers) [reply bioconda.py:52]: Why not have ```nested_dictionary_extract``` return a list? I suggest this only because I see the code currently immediately converts the generator returned from ```nested_dictionary_extract``` into a list.

File: requirements.txt
Comment (jspeed-meyers) [inline requirements.txt:4]: Why import click? I don't see it used anywhere.

**Review by jspeed-meyers — CHANGES_REQUESTED** [review]
Nice work! I left some questions and suggestions.

File: bioconda.py
Comment (spearw) [reply bioconda.py]: Done.

File: README.md
Comment (spearw) [reply README.md]: Good suggestion. Done.

File: requirements.txt
Comment (spearw) [reply requirements.txt:4]: Click is a requirement for another package - I believe zipp.

File: bioconda.py
Comment (spearw) [reply bioconda.py]: It was an oversight on my part. Good catch. Using clean_github_link.

File: test/bioconda_package_tree/amas/subfolder/meta.yaml
Comment (jspeed-meyers) [inline test/bioconda_package_tree/amas/subfolder/meta.yaml:1]: @spearw, good call on adding these other meta.yaml files that include the code substitution (jinja2 functionality). That's great and will give us all more confidence in the functionality you're building.

File: bioconda.py
Comment (spearw) [reply bioconda.py]: > Why not have `nested_dictionary_extract` return a list? I suggest this only because I see the code currently immediately converts the generator returned from `nested_dictionary_extract` into a list.

I thought that using a generator for `nested_dictionary_extract` was a little more elegant, since it isn't reliant on memory and it's a little shorter and more readable than a list version would be. Might be overkill.

**Comment by jspeed-meyers** [comment]
@spearw, I notice that the GitHub Action tests seem to be failing because of this error:

```ModuleNotFoundError: No module named 'bioconda_utils'```

Have you thought about adding that module to the requirements.txt? Or otherwise making sure the CI pipeline has access to that code? @ralatsdc might be able to help here. Also feel free to let me know if you want assistance on this point.

File: requirements.txt
Comment (jspeed-meyers) [reply requirements.txt:4]: Ahh, good reason!

File: bioconda.py
Comment (jspeed-meyers) [reply bioconda.py]: On the ```nested_dictionary_extract``` point: roger. If it works for you, it works for me. Please proceed as you propose.

**Review by ralatsdc — APPROVED** [review]

**Comment by jspeed-meyers** [comment]
And I'll approve once the tests pass! All looks good to me conceptually though.

**Comment by ralatsdc** [comment]
Tests pass here … what issue are you seeing?

**Comment by jspeed-meyers** [comment]
@ralatsdc, hmm, do you not see this pane at the bottom of the PR? Those red X's mean the tests aren't passing.
Screen Shot 2021-11-10 at 11 30 15 AM

Do you mean your tests pass locally? Or are you actually seeing a different pane on the PR? I would like to know that the tests pass in the automated CI environment.
,

**Comment by ralatsdc** [comment]
Ah. Right. This will require changing the workflow. Haven’t used these before.

In the “Install dependencies” step, we will need to
  — install Conda
  — create and activate a Conda environment for the current version of Python
  — clone the bioconda-utils repository
  — run the Conda install in the deps2repos BIOCONDA-README.md
before continuing with the remaining python installation.

Might need to activate the Conda environment in the “Test with unittest” step.

I’ll read up on the documentation, and sort it.

File: tests.py
Comment (jspeed-meyers) [inline tests.py:199]: Thanks for adding to the bioconda test suite!

File: .github/scripts/install-dependencies.sh
Comment (jspeed-meyers) [inline .github/scripts/install-dependencies.sh:1]: Nice work with this bash script.

**Review by jspeed-meyers — APPROVED** [review]
Approved. Nice work, @spearw and @ralatsdc. 👍  One question before hitting merge though: When you use this functionality and parse through ALL the bioconda recipes, does it look like you're getting a complete and accurate list of source code (e.g. GitHub) links?

File: bioconda.py
Comment (ralatsdc) [reply bioconda.py]: Note that the list may contain a single element which is a list of urls. Handled with an `isinstance()` test. Is this what is expected?

File: bioconda.py
Comment (jspeed-meyers) [inline bioconda.py:31]: Good observation. Can you please elevate the importance of this comment by placing it in the function docstrings prominently above?

File: bioconda.py
Comment (jspeed-meyers) [inline bioconda.py:44]: Technically the links are not necessarily to GitHub. They could be to GitLab, for instance. Hmmm. Perhaps we need a little re-engineering. Perhaps we discuss tomorrow.

File: bioconda.py
Comment (jspeed-meyers) [inline bioconda.py:63]: Could you please explain a bit more here please?

File: test/bioconda_package_tree/abacas/patches/path.patch
Comment (jspeed-meyers) [inline test/bioconda_package_tree/abacas/patches/path.patch:1]: What is this thing?

File: test/bioconda_package_tree/abundancebin/0001-Use-CXX-and-flags.patch
Comment (jspeed-meyers) [inline test/bioconda_package_tree/abundancebin/0001-Use-CXX-and-flags.patch:1]: Same here: What is this thing?

File: test/bioconda_package_tree/abundancebin/build.sh
Comment (jspeed-meyers) [inline test/bioconda_package_tree/abundancebin/build.sh:1]: ?

File: test/bioconda_package_tree/admixture/build.sh
Comment (jspeed-meyers) [inline test/bioconda_package_tree/admixture/build.sh:1]: ?

File: utils.py
Comment (jspeed-meyers) [inline utils.py:74]: Have you thought about using (arguably) slightly simpler approaches like the ones in this SO post? 

https://stackoverflow.com/questions/952914/how-to-make-a-flat-list-out-of-a-list-of-lists

**Review by jspeed-meyers — COMMENTED** [review]
@spearw and @ralatsdc, a few more questions and comments. Nice work.

**Pull request merged**
